org.eclipse.xtext.web/src/main/java/org/eclipse/xtext/web/server/formatting/FormattingService.java

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright (c) 2015, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.web.server.formatting;
+
+import java.util.List;
+
+import org.eclipse.xtext.formatting.INodeModelFormatter;
+import org.eclipse.xtext.formatting2.FormatterRequest;
+import org.eclipse.xtext.formatting2.IFormatter2;
+import org.eclipse.xtext.formatting2.regionaccess.ITextRegionAccess;
+import org.eclipse.xtext.formatting2.regionaccess.ITextReplacement;
+import org.eclipse.xtext.formatting2.regionaccess.TextRegionAccessBuilder;
+import org.eclipse.xtext.formatting2.regionaccess.internal.TextSegment;
+import org.eclipse.xtext.nodemodel.ICompositeNode;
+import org.eclipse.xtext.parser.IParseResult;
+import org.eclipse.xtext.preferences.ITypedPreferenceValues;
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.util.ITextRegion;
+import org.eclipse.xtext.util.TextRegion;
+import org.eclipse.xtext.util.Wrapper;
+import org.eclipse.xtext.util.concurrent.CancelableUnitOfWork;
+import org.eclipse.xtext.web.server.InvalidRequestException;
+import org.eclipse.xtext.web.server.model.IXtextWebDocument;
+import org.eclipse.xtext.web.server.model.XtextWebDocumentAccess;
+
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import com.google.inject.Singleton;
+
+/**
+ * Service class for text formatting.
+ */
+@Singleton
+public class FormattingService {
+	@Inject(optional = true)
+	private INodeModelFormatter formatter1;
+
+	@Inject(optional = true)
+	private Provider<IFormatter2> formatter2Provider;
+
+	@Inject
+	private Provider<FormatterRequest> formatterRequestProvider;
+
+	@Inject
+	private TextRegionAccessBuilder regionBuilder;
+
+	/**
+	 * Format the given document. This operation modifies the document content and
+	 * returns the
+	 */
+	public FormattingResult format(XtextWebDocumentAccess document, ITextRegion selection,
+			ITypedPreferenceValues preferences) throws InvalidRequestException {
+		Wrapper<String> textWrapper = new Wrapper<String>();
```

Comment:
Might be useful to have a preconditions checking that at one of formatter1 or formatter2Provider is not null.

There is already an IllegalStateException in the else branch but that isn't as clear